mvlearn/factorization/ajive.py

```diff
@@ -0,0 +1,921 @@
+    if precomp_init_svd is None:
+        precomp_init_svd = {bn: None for bn in block_names}
+    precomp_init_svd = _dict_formatting(precomp_init_svd)
+    assert set(precomp_init_svd.keys()) == set(block_names)
+    for bn in block_names:
+        udv = precomp_init_svd[bn]
+        if udv is not None and not hasattr(udv, "keys"):
+            precomp_init_svd[bn] = {
+                "scores": udv[0],
+                "svals": udv[1],
+                "loadings": udv[2],
+            }
+
+    # Check initial signal ranks
+    if precomp_init_svd is None:
+        precomp_init_svd = {bn: None for bn in block_names}
+    init_signal_ranks = _dict_formatting(init_signal_ranks)
+    assert set(init_signal_ranks.keys()) == set(block_names)
+
+    # signal rank must be at least one lower than the shape of the block
+    for bn in block_names:
+        assert 1 <= init_signal_ranks[bn]
+        assert init_signal_ranks[bn] <= min(blocks[bn].shape) - 1
+
+    # Check the joint ranks
+    if joint_rank is not None and joint_rank > sum(init_signal_ranks.values()):
+        raise ValueError(
+            "joint_rank must be smaller than the sum of the initial signal \
+            ranks"
+        )
+
+    # Check individual ranks
+    if indiv_ranks is None:
+        indiv_ranks = {bn: None for bn in block_names}
+    indiv_ranks = _dict_formatting(indiv_ranks)
+    assert set(indiv_ranks.keys()) == set(block_names)
+
+    for k in indiv_ranks.keys():
+        assert indiv_ranks[k] is None or type(indiv_ranks[k]) in [int, float]
+
+    # Check centering
+    if type(center) == bool:
+        center = {bn: center for bn in block_names}
+    center = _dict_formatting(center)
+
+    return (
+        blocks,
+        init_signal_ranks,
+        indiv_ranks,
+        precomp_init_svd,
+        center,
+        obs_names,
+        var_names,
+        shapes,
+    )
+
+
+def _reconsider_joint_components(
+    blocks, sv_threshold, joint_scores, joint_svals, joint_loadings, joint_rank
+):
+    """
+    Checks the identifiability constraint on the joint singular values
+    """
+
+    # check identifiability constraint
+    to_keep = set(range(joint_rank))
+    for bn in blocks.keys():
+        for j in range(joint_rank):
+            # This might be joint_sv
+            score = np.dot(blocks[bn].T, joint_scores[:, j])
+            sv = np.linalg.norm(score)
+
+            # if sv is below the threshold for any data block remove j
+            if sv < sv_threshold[bn]:
+                print("removing column " + str(j))
+                to_keep.remove(j)
+                break
+
+    # remove columns of joint_scores that don't satisfy the constraint
+    joint_rank = len(to_keep)
+    joint_scores = joint_scores[:, list(to_keep)]
+    joint_loadings = joint_loadings[:, list(to_keep)]
+    joint_svals = joint_svals[list(to_keep)]
+    return joint_scores, joint_svals, joint_loadings, joint_rank
+
+
+class ViewSpecificResults(object):
```

Comment:
I think this makes more sense in the utils folder? It is never passed to the user, is just used for internal storage. It's going to appear in the webpage if here.

Reply:
I didn't create a .rst module for it so when I built the HTML it didn't seem to be there. Do you think that it could still show up when merged? Either way, not a big deal to move to utils

Reply:
Moved it to the pca file. Doesn't work in the utils file bc the pca file calls it (becomes circular)